### PR TITLE
feat/save-to-mydrive

### PR DESCRIPTION
**Pull request by maya-fisher**

### CODE REVIEW

**Review by sternomer — APPROVED** [review]

File: file/file.go [hunk 1]
```diff
@@ -134,6 +140,10 @@ const (
 
 	// FileIDIsRequiredMessage is the error message for missing fileID
 	FileIDIsRequiredMessage = "fileID is required"
+
+	// CopyFileRole is the role that is required of the authenticated requester to have to be
+	// permitted to make the CopyFile action.
+	CopyFileRole = ppb.Role_READ
```
Comment (yoavrozner) [inline file/file.go:146]: Please check

File: file/file.go [hunk 2]
```diff
@@ -1466,3 +1477,126 @@ func stringInSlice(a string, list []string) bool {
...
+	}
+
+	folderID := c.Param(ParamFolderID)
+	if folderID == "undefined" {
+		folderID = ""
+	}
+
+	newFileName := c.Param(ParamNewFileName)
+	if newFileName == "undefined" {
+		newFileName = ""
```
Comment (yoavrozner) [inline file/file.go:1504]: Reeturn error!

File: file/file.go [hunk 3]
```diff
@@ -1466,3 +1477,126 @@ func stringInSlice(a string, list []string) bool {
...
+	}
+
+
+	userFilePermission, _ := r.HandleUserFilePermission(c, fileID, CopyFileRole)
+	if userFilePermission == "" {
+		if !r.HandleUserFilePermit(c, fileID, CopyFileRole) {
+			c.AbortWithStatus(http.StatusUnauthorized)
+			return
+		}
+	}
```
Comment (yoavrozner) [inline file/file.go:1514]: Triple check

File: file/file.go [hunk 4]
```diff
@@ -1466,3 +1477,126 @@ func stringInSlice(a string, list []string) bool {
...
+		Parent:  folderID,
+		AppID:   fileToCopy.AppID,
+	})
+
+	if err != nil {
+		httpStatusCode := gwruntime.HTTPStatusFromCode(status.Code(err))
+		loggermiddleware.LogError(r.logger, c.AbortWithError(httpStatusCode, err))
+
+		return
+	}
```
Comment (yoavrozner) [inline file/file.go:1544]: Check it out Rollback

File: file/file.go [hunk 5]
```diff
@@ -1466,3 +1477,126 @@ func stringInSlice(a string, list []string) bool {
...
+	if err != nil {
+		return
+	}
+
+	if err != nil {
+		httpStatusCode := gwruntime.HTTPStatusFromCode(status.Code(err))
+		loggermiddleware.LogError(r.logger, c.AbortWithError(httpStatusCode, err))
+
+		return
+	}
```
Comment (yoavrozner) [inline file/file.go:1571]: What?

File: file/file.go [hunk 6]
```diff
@@ -1466,3 +1477,126 @@ func stringInSlice(a string, list []string) bool {
...
+		KeySrc:     fileToCopy.Key,
+		KeyDest:    key,
+	})
+
+	if err != nil {
+		httpStatusCode := gwruntime.HTTPStatusFromCode(status.Code(err))
+		loggermiddleware.LogError(r.logger, c.AbortWithError(httpStatusCode, err))
+
+		return
+	}
```
Comment (yoavrozner) [inline file/file.go:1585]: Rollback

File: file/file.go [hunk 7]
```diff
@@ -1466,3 +1477,126 @@ func stringInSlice(a string, list []string) bool {
...
+	if err != nil {
+		httpStatusCode := gwruntime.HTTPStatusFromCode(status.Code(err))
+		loggermiddleware.LogError(r.logger, c.AbortWithError(httpStatusCode, err))
+		return ""
+	}
+
+	key := keyResp.GetKey()
+
+	return key
+}
```
Comment (yoavrozner) [inline file/file.go:1602]: Utils please

**Review by yoavrozner — CHANGES_REQUESTED** [review]
Well done!

File: file/file.go (hunk 1 above)
Comment (maya-fisher) [reply file/file.go:146]: It's the same as in favorite

File: file/file.go (hunk 2 above)
Comment (maya-fisher) [reply file/file.go:1504]: Done

File: file/file.go (hunk 5 above)
Comment (maya-fisher) [reply file/file.go:1571]: Done

File: file/file.go (hunk 6 above)
Comment (maya-fisher) [reply file/file.go:1585]: Done. but need to check

File: file/file.go (hunk 4 above)
Comment (maya-fisher) [reply file/file.go:1544]: Done, but need to check

**Review by yoavrozner — DISMISSED** [review]
Good job @maya-fisher

File: file/file.go [hunk 8]
```diff
@@ -1466,3 +1477,125 @@ func stringInSlice(a string, list []string) bool {
...
+		return
+	}
+
+	folderID := c.Param(ParamFolderID)
+	if folderID == "undefined" {
+		folderID = ""
+	}
+
+	newFileName := c.Param(ParamNewFileName)
+	if newFileName == "undefined" {
```
Comment (yoavrozner) [inline file/file.go:1503]:
```suggestion
	if newFileName == "undefined" || newFileName == "" {
```

**Review by yarinBenisty — APPROVED** [review]

**Pull request merged**